### PR TITLE
RavenDB-18008 - [Sharding] Allow to get database stats (shard wide)

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18008

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio

### CODE REVIEW

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                    writer.WriteStartObject();
+
+                    WriteDatabaseStatisticsInternal(writer, statistics);
+
+                    writer.WriteEndObject();
+                }
+            }
+        }
+
+        private static void WriteDatabaseStatisticsInternal(AbstractBlittableJsonTextWriter writer, DatabaseStatistics statistics)
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: Can we reuse the code from `BlittableJsonTextWriterExtensions`?

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: done

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                        combined.CountOfIndexes = result.CountOfIndexes;
+                }
+
+                combined.SizeOnDisk = new Size(totalSizeOnDisk);
+                combined.TempBuffersSizeOnDisk = new Size(totalTempBuffersSizeOnDisk);
+
+                return combined;
+            }
+
+            public RavenCommand<DatabaseStatistics> CreateCommandForShard(int shard) => new GetStatisticsOperation.GetStatisticsCommand(null, null);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: what are those `null`,`null`? please use the named parameters
e.g. `GetStatisticsCommand(foo: null, bar: null)`

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                    combined.CountOfCounterEntries += result.CountOfCounterEntries;
+                    combined.CountOfDocumentsConflicts += result.CountOfDocumentsConflicts;
+                    combined.CountOfRevisionDocuments += result.CountOfRevisionDocuments;
+                    combined.CountOfTimeSeriesSegments += result.CountOfTimeSeriesSegments;
+                    combined.CountOfTombstones += result.CountOfTombstones;
+                    totalSizeOnDisk += result.SizeOnDisk.SizeInBytes;
+                    totalTempBuffersSizeOnDisk += result.TempBuffersSizeOnDisk.SizeInBytes;
+
+                    if (combined.CountOfIndexes == -1)
+                        combined.CountOfIndexes = result.CountOfIndexes;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: let's get the indexes & the index count from the database record if possible

**Review by karmeli87 — CHANGES_REQUESTED** [review]

**Comment by karmeli87** [comment]
Please add a simple test for this as well.

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: done

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: done

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs [hunk 4]
```diff
@@ -28,6 +31,33 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+            foreach (var key in indexes.Keys)
+            {
+                var index = indexes[key];
+
+                indexInformation[i] = new IndexInformation
+                {
+                    State = index.State ?? IndexState.Error,
+                    Name = index.Name,
+                    LockMode = index.LockMode ?? IndexLockMode.LockedError,
+                    Priority = index.Priority ?? new IndexPriority(),
```
Comment (karmeli87) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: `IndexPriority` is an `enum` so it is wired for me to see `new IndexPriority()`.
Does the `Priority`/`LockMode`/`State` are nullable?

Wonder what about `IsStale`.. not sure what to do with that, but let's set it explicitly here.

Also the `State` is from the index definition while the actual state can be different. (e.g. we can disable an index locally) - We don't want to handle this right now, but do need to keep that in mind

File: src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream()))
+                {
+                    Json.BlittableJsonTextWriterExtensions.WriteDatabaseStatistics(writer, null, statistics);
+                }
+            }
+        }
+
+        private IndexInformation[] GetDatabaseIndexesFromRecord()
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedStatsHandler.cs]: Wouldn't it be better to move this inside the ShardedStatsOperation? Then each time I execute this (e.g. in a different place than this endpoint) the results will be consistent. cc @karmeli87 Thoughts?

File: src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        where TOperationContext : JsonOperationContext
+    {
+        protected AbstractStatsHandlerProcessor([NotNull] TRequestHandler requestHandler, [NotNull] JsonContextPoolBase<TOperationContext> contextPool) : base(requestHandler, contextPool)
+        {
+        }
+
+        protected IndexInformation[] GetDatabaseIndexesFromRecord()
+        {
+            if (RequestHandler is ShardedRequestHandler shardedRequestHandler)
+            {
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessor.cs]: the sharded probably should simply override this?

Comment (karmeli87) [reply src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessor.cs]: actually `GetDatabaseIndexesFromRecord` should be only at the sharded, no?

Comment (karmeli87) [reply src/Raven.Server/Documents/Handlers/Processors/AbstractStatsHandlerProcessor.cs]: okay, I realize there is a problem with the inheritance and we can't have a common abstract class..
wouldn't it be better to hook such things on the `ShardedContext` ? an extension for the  `record.Indexes` ?

**Review by ppekrol — APPROVED** [review]

**Pull request merged**